File: java-to-go-projects/09-markdown-to-html/Main.java
package markdown;

import java.nio.file.*;

/**
 * CLI entry point for the Markdown-to-HTML converter.
 *
 * Usage: java Main input.md [output.html]
 *
 * Go structure:
 *   main.go               — CLI, file I/O
 *   markdown/token.go     — Token struct + Type constants
 *   markdown/lexer.go     — Tokenize(source) using regexp
 *   markdown/renderer.go  — Render(tokens) → HTML string
 *                            processInline for bold/italic/code/link
 *
 * Rust structure:
 *   main.rs               — CLI, file I/O
 *   token.rs              — Token enum with variants (Heading{level,content}, etc.)
 *   lexer.rs              — tokenize(source) using regex crate
 *   renderer.rs           — render(tokens) → String, process_inline()
 *
 * Key learning:
 *   - Go: regexp package, string manipulation, iota for constants
 *   - Rust: enum variants with data, regex captures, String ownership in nested transforms
 */
public class Main {

    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.err.println("Usage: md2html <input.md> [output.html]");
            System.exit(1);
        }

        String source = Files.readString(Path.of(args[0]));

        Lexer lexer = new Lexer();
        Renderer renderer = new Renderer();

        var tokens = lexer.tokenize(source);
        String html = renderer.render(tokens);

        if (args.length >= 2) {
            Files.writeString(Path.of(args[1]), html);
            System.out.println("Written to " + args[1]);
        } else {
            System.out.println(html);
        }
    }
}

File: java-to-go-projects/09-markdown-to-html/go/main.go
package main

import (
	"fmt"
	"os"
	"regexp"
	"strings"
)

const (
	TokenHeading = iota
	TokenParagraph
	TokenCodeBlock
	TokenListItem
	TokenOrderedItem
	TokenHorizontalRule
	TokenBlockquote
	TokenNewline
)

type Token struct {
	Type    int
	Content string
	Level   int    // for headings
	URL     string // for links
}

// tokenize converts raw Markdown source into a slice of block-level Tokens.
func tokenize(source string) []Token {
	_ = regexp.MustCompile
	_ = strings.Split(source, "\n")
	panic("todo")
}

// render converts a slice of Tokens into a complete HTML document string.
func render(tokens []Token) string {
	var sb strings.Builder
	_ = &sb
	panic("todo")
}

// processInline applies inline formatting (bold, italic, code, links) to text.
func processInline(text string) string {
	panic("todo")
}

// escapeHTML escapes <, >, &, " characters.
func escapeHTML(text string) string {
	panic("todo")
}

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s <input.md> [output.html]\n", os.Args[0])
		os.Exit(1)
	}

	panic("todo: read input, tokenize, render, write output")
}
